src/main/java/org/rumbledb/runtime/control/TypeswitchRuntimeIterator.java

```diff
@@ -114,6 +126,7 @@ private boolean testTypeMatch(TypeswitchRuntimeIteratorCase typeSwitchCase) {
                 }
             }
         }
+        this.matchingIterator = null;
```

Comment:
I think this was causing the issue

Reply:
This is a very nice catch.

I will discuss an orthogonal issue to what you have fixed here. I believe this function does 'too many' things in contrast to what its name implies. Lines type checks performed in 112-114 alongside returning true or false indeed perform a type check. But this function modifies state by adding variables and updating the matching iterator as well. This makes it a very likely spot to create future bugs. For example, a developer might instinctively move the code at line 89:
`this.testValue = test.materializeFirstItemOrNull(this.currentDynamicContextForLocalExecution);` after the loop that's performing the testTypeMatch() to increase locality of the testValue variable by bringing it closer to where it's used. This could create some nasty bugs due to the obfuscated modification of the state that alters the dynamic context.

I would recommend only keeping the lines I mentioned earlier in this function to make sure it 'only' performs a type matching. And the modification of the state in lines 115-199: 
```
if (typeSwitchCase.getVariableName() != null) {
    ....
}
 this.matchingIterator = typeSwitchCase.getReturnIterator(); 
``` 
as well as the line you added at 126 `this.matchingIterator = null;` is taken out of this function. They could either be directly placed within the loop in initializeIterator or extracted to a separate function.

Reply:
Done, I have made the iterator the return value, so that initializeIterator is now responsible for modifying the state, and it is now handling cases and the default case with similar code.